### PR TITLE
FEXCore: Support preserve_all ABI for interpreter fallbacks

### PR DESCRIPTION
**Pull request by Sonicadvance1**
This significantly improves performance of x87 instructions when operating in full precision mode. Resulting performance improvement can be seen from microbenchmark results.
Image_2023-09-18_17-20-06

Additionally outside of microbenchmark, this improves Linux native Psychonaut's title screen from 34 FPS to 51FPS. A perfect 50% additional performance.
This doesn't improve performance to the levels of the fp64 precision x87 mode but it is a decent uplift for the default option.

Additionally this `preserve_all` ABI path is only available on clang-17 which isn't out yet. So no one will be able to see the performance improvements yet.

### CODE REVIEW

File: FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp [hunk 1]
```diff
@@ -144,6 +236,101 @@ namespace x32 {
     FEXCore::ARMEmitter::VReg::v28, FEXCore::ARMEmitter::VReg::v29,
     FEXCore::ARMEmitter::VReg::v30, FEXCore::ARMEmitter::VReg::v31
   };
+
```
Comment (alyssarosenzweig) [inline FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp:239]: Can we deduplicate this chunk of code? failing to deduplicate stuff in FEX has bit me several times. I don't see why this instance should be different

File: FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp [hunk 2]
```diff
@@ -532,111 +719,241 @@ void Arm64Emitter::FillStaticRegs(bool FPRs, uint32_t GPRFillMask, uint32_t FPRF
...
+    // rsp capable move
+    add(ARMEmitter::Size::i64Bit, TmpReg, ARMEmitter::Reg::rsp, 0);
+
+    // Push the vector registers
+    PushVectorRegisters(TmpReg, CanUseSVE, DynamicFPRs);
+
+    // Push the general registers.
+    PushGeneralRegisters(TmpReg, DynamicGPRs);
+  }
+  else {
```
Comment (alyssarosenzweig) [inline FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp:890]: surely we can/should dedupe the two sides of the if?

File: FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp [hunk 3]
```diff
@@ -532,111 +719,241 @@ void Arm64Emitter::FillStaticRegs(bool FPRs, uint32_t GPRFillMask, uint32_t FPRF
...
+      DynamicFPRs = x32::PreserveAll_DynamicFPRSVE;
+    }
+
+    // Fill static registers second.
+    FillStaticRegs(true, x32::PreserveAll_SRAMask, CanUseSVE ? x32::PreserveAll_SRAFPRSVEMask : x32::PreserveAll_SRAFPRMask);
+
+    // Pop vectors first
+    PopVectorRegisters(CanUseSVE, DynamicFPRs);
+
+    // Pop GPRs second
```
Comment (alyssarosenzweig) [inline FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp:952]: ditto. also, the comments are out of sync with the code

File: FEXCore/Source/Interface/Core/JIT/Arm64/JIT.cpp [hunk 4]
```diff
@@ -84,9 +84,14 @@ void Arm64JITCore::Op_Unhandled(IR::IROp_Header const *IROp, IR::NodeID Node) {
   } else {
     switch(Info.ABI) {
       case FABI_F80_I16_F32:{
-        SpillStaticRegs(TMP1);
+        if (Info.SupportsPreserveAllABI) {
```
Comment (alyssarosenzweig) [inline FEXCore/Source/Interface/Core/JIT/Arm64/JIT.cpp:87]: probably want to factor out a helper for this

File: FEXCore/Source/Interface/Core/JIT/Arm64/JIT.cpp [hunk 5]
```diff
@@ -97,9 +102,13 @@ void Arm64JITCore::Op_Unhandled(IR::IROp_Header const *IROp, IR::NodeID Node) {
         blr(ARMEmitter::Reg::r1);
 #endif
 
-        PopDynamicRegsAndLR();
-
-        FillStaticRegs();
+        if (Info.SupportsPreserveAllABI) {
```
Comment (alyssarosenzweig) [inline FEXCore/Source/Interface/Core/JIT/Arm64/JIT.cpp:105]: and this

File: FEXCore/Source/Interface/Core/Interpreter/Fallbacks/F80Fallbacks.h [hunk 6]
```diff
@@ -141,14 +155,33 @@ struct OpHandlers<IR::OP_F80CVTTOINT> {
 
 template<>
 struct OpHandlers<IR::OP_F80ROUND> {
+  FEXCORE_PRESERVE_ALL_ATTR
   static X80SoftFloat handle(uint16_t NewFCW, X80SoftFloat Src1) {
-    LoadDeferredFCW(NewFCW);
```
Comment (alyssarosenzweig) [inline FEXCore/Source/Interface/Core/Interpreter/Fallbacks/F80Fallbacks.h:145]: unrelated change?

**Review by alyssarosenzweig — CHANGES_REQUESTED** [review]

File: FEXCore/Source/Interface/Core/Interpreter/Fallbacks/F80Fallbacks.h (hunk 6 above)
Comment (Sonicadvance1) [reply FEXCore/Source/Interface/Core/Interpreter/Fallbacks/F80Fallbacks.h:145]: Removed

File: FEXCore/Source/Interface/Core/JIT/Arm64/JIT.cpp (hunk 5 above)
Comment (Sonicadvance1) [reply FEXCore/Source/Interface/Core/JIT/Arm64/JIT.cpp]: Added helper

File: FEXCore/Source/Interface/Core/JIT/Arm64/JIT.cpp (hunk 4 above)
Comment (Sonicadvance1) [reply FEXCore/Source/Interface/Core/JIT/Arm64/JIT.cpp]: Added helper

File: FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp (hunk 2 above)
Comment (Sonicadvance1) [reply FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp:879]: Deduplicated

File: FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp (hunk 3 above)
Comment (Sonicadvance1) [reply FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp]: Deduplicated

File: FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp (hunk 1 above)
Comment (Sonicadvance1) [reply FEXCore/Source/Interface/Core/ArchHelpers/Arm64Emitter.cpp:239]: I have a comment there lamenting that I wish we could constexpr to deduplicate but we lack the library/compiler support for now. We could potentially switch over to parsing a uint32_t mask but that would require refactoring/duplicating a bunch of the push/pop routines to support one or the other.

**Comment by neobrain** [comment]
Nice results!

Maybe I missed some bit of documentation in the diff that explains this, but what does this actually do and how does it help improve x87 performance?

**Comment by alyssarosenzweig** [comment]
> Nice results!
> 
> Maybe I missed some bit of documentation in the diff that explains this, but what does this actually do and how does it help improve x87 performance?

Right now, FEX spills/fills EVERYTHING whenever it calls out to a fallback, due to the ABI mismatch between C code and FEX's SRA. This means x87 code is slooooooooow since every nontrivial x87 instruction is a fallback. This PR changes the calling convention to avoid needing to spill so many registers across call sites, instead requiring the C code to spill just the few registers it actually needs. ==> many fewer instructions executed overall

**Review by alyssarosenzweig — APPROVED** [review]
